src/sidebar/store/modules/drafts.js

```diff
@@ -0,0 +1,175 @@
+'use strict';
+
+const util = require('../util');
+
+/**
+ * The drafts store provides temporary storage for unsaved edits to new or
+ * existing annotations.
+ *
+ * A draft consists of:
+ *
+ * 1. `model` which is the original annotation domain model object which the
+ *    draft is associated with. Domain model objects are never returned from
+ *    the drafts service, they're only used to identify the correct draft to
+ *    return.
+ *
+ * 2. `isPrivate` (boolean), `tags` (array of objects) and `text` (string)
+ *    which are the user's draft changes to the annotation. These are returned
+ *    from the drafts service by `drafts.getDraft()`.
+ *
+ */
+function init() {
+  return {
+    drafts: [],
+  };
+}
+
+/* Reducer */
+
+const update = {
+  DISCARD_ALL_DRAFTS: function(state, action) {
+    return {
+      drafts: []
+    }
+  },
+  REMOVE_DRAFT: function(state, action) {
+    const drafts = state.drafts.filter(draft => {
+      return !draft.match(action.model);
+    });
+    return {
+      drafts
+    };
+  },
+  ADD_DRAFT: function(state, action) {
+    return { 
+      drafts: state.drafts.concat(action.draft) // non-mutating
+    };
+  },
+};
+
+/* Actions */
+
+/**
+ * Create or update the draft version for a given annotation by
+ * replacing any existing draft or simply creating a new one.
+ */
+function createDraft(model, changes) {
+  return function(dispatch, getState) {
+    dispatch(removeDraft(model));
+    const draft = new Draft(model, changes);
+    dispatch({
+      type: actions.ADD_DRAFT,
+      draft
+    });
+  }
+}
+
+/** Remove all drafts. */
+function discardAllDrafts() {
+  return { 
+    type: actions.DISCARD_ALL_DRAFTS,
+  };
+}
+
+/** Remove the draft version of an annotation. */
+function removeDraft(model) {
+  return { 
+    type: actions.REMOVE_DRAFT, 
+    model: model 
+  };
+}
+
+const actions = util.actionTypes(update);
+
+/* Selectors */
+
+/**
+ * Returns the number of drafts - both unsaved new annotations, and unsaved
+ * edits to saved annotations - currently stored.
+ */
+function countDrafts(state) {
+  return state.drafts.length;
+}
+
+/** Retrieve the draft changes for an annotation. */
+function getDraft(state, model) {
+  for (let i = 0; i < state.drafts.length; i++) {
+    const draft = state.drafts[i];
+    if (draft.match(model)) {
+      return draft;
+    }
+  }
+  return null;
+}
+
+/**
+ * Returns the draft changes for an annotation, or null if no draft exists
+ * or the draft is empty.
+ */
+function getDraftIfNotEmpty(state, model) {
+  const draft = getDraft(state, model);
+  if (!draft) {
+    return null;
+  }
+  return draft.isEmpty() ? null : draft;
```

Comment:
What cases would the draft exist but be empty? Is that when you open an annotation for editing but haven't made any changes?

Reply:
Yes. IIRC when you click on an annotation to start editing it, the initial state is that you have a draft with no changes. There is no separate "is editing" flag.